### PR TITLE
feat: add Visual JSON Schema Editor

### LINKED ISSUE #755: Visual Designer: Message schema
Add the schema into the design system

Screenshot 2023-07-31 at 15 09 13

@fmvilas already did it in the past - we can reuse some of his stuffs here.  

Related to #748

### PR DESCRIPTION
**Pull request by princerajpoot20**
**Adding Visual JSON Schema Editor**

Resolves: #755

### CODE REVIEW

File: packages/ui/components/CodeEditor.tsx
Comment (fmvilas) [inline packages/ui/components/CodeEditor.tsx:10]:
```suggestion
interface CodeEditorProps {
  schema: SchemaObject;
  onChange: (newSchema: SchemaObject) => void;
}

export const CodeEditor: React.FC<CodeEditorProps> = ({ schema, onChange }) => {
```

A couple of things:
1. You have the SchemaObject type [here](https://github.com/asyncapi/parser-js/blob/master/src/spec-types/v3.ts#L401) and [here](https://github.com/asyncapi/parser-js/blob/master/src/spec-types/v2.ts#L227). You may want to propose that we extract them in a separate package so you don't depend on parser here.
2. I'd get rid of `Schema` in `onSchemaChange` because it's redundant.

File: packages/ui/components/VisualEditor.tsx
Comment (fmvilas) [inline packages/ui/components/VisualEditor.tsx:141]:
```suggestion
      onRemove(path);
```

File: packages/ui/components/VisualEditor.tsx
Comment (fmvilas) [inline packages/ui/components/VisualEditor.tsx:155]:
```suggestion
    onToggleRequired(path);
```

File: packages/ui/components/VisualEditor.tsx
Comment (fmvilas) [inline packages/ui/components/VisualEditor.tsx:180]: Same as with CodeEditor component:

```suggestion
interface VisualEditorProps {
  schema: SchemaObject;
  onChange: (updatedSchema: SchemaObject) => void;
}

export const VisualEditor: React.FC<VisualEditorProps> = ({ schema, onChange }) => {
```

File: packages/ui/components/index.tsx
Comment (fmvilas) [inline packages/ui/components/index.tsx:17]: Since this component is going to be the code editor of the visual editor, I would not export it directly here but as a subcomponent of VisualEditor. Just to avoid conflicting with the generic CodeEditor that we'll have in the future.

**Review by fmvilas — CHANGES_REQUESTED** [review]
Dude that was fast 👏

**Comment by aeworxet** [comment]
@asyncapi/bounty_team

**Comment by aeworxet** [comment]
@princerajpoot20
Please edit the description of the PR if it's meant to resolve the Bounty Issue you're working on (so it's displayed in the issue's comments and GitHub search, with 'Resolves *full_URL_to_GH_issue*').

File: packages/ui/components/CodeEditor.tsx
Comment (princerajpoot20) [reply packages/ui/components/CodeEditor.tsx]: @fmvilas Could you explain this further. Is it like extracting a `SchemaObject` from the `parser` into a separate file and using this as a type instead of using its string type?

File: packages/ui/components/index.tsx
Comment (princerajpoot20) [reply packages/ui/components/index.tsx:17]: sure. 👍

File: packages/ui/components/CodeEditor.tsx
Comment (fmvilas) [reply packages/ui/components/CodeEditor.tsx]: Yes, that's exactly what I meant 👍

**Comment by princerajpoot20** [comment]
### Update

- [x] The "add property" option is now working for inner objects as well.

What's left:

- [ ] Completing the Example section.
- [ ] UI changes.

**Comment by princerajpoot20** [comment]
### Currently working on:
I am working on resolving type error that is causing the build to fail, though it runs on my local machine.

**Comment by aeworxet** [comment]
Is something blocking this PR from merging, though the functionality is not finished, but [agreed](https://github.com/asyncapi/studio/issues/755#issuecomment-1999207498) to be merged 'as is?'
Is Fran's [changerequest](https://github.com/asyncapi/studio/pull/905#pullrequestreview-1788557119) still relevant (it's dated 2023-12-19?)

**Review by Amzani — DISMISSED** [review]
LGTM

**Comment by Amzani** [comment]
@princerajpoot20 could you check @fmvilas  suggestion: https://github.com/asyncapi/studio/pull/905#discussion_r1431220019 if it's not relevant anymore let him know so he can approve as well

We can merge "as is" and make a new iteration to adjust it.

**Review by fmvilas — CHANGES_REQUESTED** [review]
I'm confused. I'm trying to use the component in the design system (in Storybook) and it doesn't work when I try to remove fields or mark them as required. Is it a thing of how we're integrating it in Storybook or is it a thing of the component? Would love to see it working completely as if it was in the real app.

Also, the design still needs some work. We should make it match the proposed design IMHO.

That said, I'm not a code owner of Studio. If code owners think we should merge this, I'm happy to dismiss my review. That's totally fine, we can address these issues later down the track. Just wanted to highlight the things I'm seeing.

**Review by fmvilas — APPROVED** [review]
Just got a DM from @princerajpoot20 explaining the situation:

> I’ve discussed this with Samir, and we’ve decided to complete it in two iterations.
> We’ve adjusted the complexity of the bounty from advanced to [two medium bounties](https://github.com/asyncapi/studio/issues/755#issuecomment-1999335416).
> I have completed the first iteration, and the second iteration is scheduled for Q2 of 2024 ([Link for 2nd iteration](https://github.com/asyncapi/studio/issues/1023)).
> This iteration will include all remaining tasks, such as deletion, toggling between required/not required, and UI changes. Hence we are merging the current progress now and will complete the remainder in the second iteration.

Approving! 👍

**Review by Amzani — APPROVED** [review]

**Pull request merged**
